### PR TITLE
Wait-free size for lock-free data structures

### PR DESCRIPTION
**Pull request by polytypic**
This PR implements a wait-free size counter for lock-free data structures.

This is a new algorithm inspired by the paper [Concurrent Size](https://arxiv.org/pdf/2209.07100.pdf) by Gal Sela and Erez Petrank.  This differs from the paper in some important ways.  In particular, unlike in the paper, the new algorithm does not require the number of threads to be limited and given unique integer indices to ensure correctness.  Instead, the new algorithm uses a lock-free transactional approach to performing the counter updates at most once.  The new algorithm is also designed to give correct answers in case of internal counter overflow.

A counter contains an array of atomics.  During the development of this I noticed that OCaml 5 generates inefficient code for accessing an array of atomics, because the compiler fails to notice that an atomic cannot be a `float`.  To avoid the inefficiency and to also work around the CSE issues of `Atomic.get`, this implementation uses a `Transparent_atomic` module from the `multicore-magic` library, which is a replacement for `Stdlib.Atomic` that works around both of these issues (incorrect CSE optimization and array of atomics performance).  As always, the hope is that in the future `multicore-magic` may be retired as obsolete.

Note that this depends on two PRs to multicore-magic:
* [Add `Transparent_atomic` module](https://github.com/ocaml-multicore/multicore-magic/pull/8)
* [Add `instantaneous_domain_index`](https://github.com/ocaml-multicore/multicore-magic/pull/10)

---

I've been experimenting with some benchmarks and the allocation of a new `once` object for each operation seems to be relatively expensive.  With each `once` freshly allocated one can get up to 63 M ops/s (on a single core).  Reusing `once` objects (in an unrealistic manner) one can get up to 195 M ops/s (on a single core).  At any rate, I suspect it is possible to do better than the 63 M ops/s that the code in this PR currently achieves (on a single core of my laptop).

---

We should have a proper benchmark for this &mdash; and I actually have one &mdash; but I'll leave that to a future PR.

### CODE REVIEW

File: saturn_lockfree.opam [hunk 1]
```diff
@@ -34,3 +35,6 @@ build: [
   ]
 ]
 dev-repo: "git+https://github.com/ocaml-multicore/saturn.git"
+pin-depends: [
+  [ "multicore-magic.dev" "git+https://github.com/ocaml-multicore/multicore-magic#e7abfa84e0657d7d56f977a5695f5c20b1f465a1" ]
```
Comment (polytypic) [inline saturn_lockfree.opam:39]: This pin-depend needs to eliminated before merging.

**Comment by lyrm** [comment]
So, if I am getting this right, a node is semantically added not when a connection to it is made, but when its `incr` counter is used for the first time, which should be either : 
- at the end of the `add` operation 
- or when anyone else witness the new node for the first time, depending on which happens first

 So the first call of `Size.update_once` on `node.incr` is the point of lineaarization of `add t node` ?

**Comment by polytypic** [comment]
> a node is semantically added not when a connection to it is made, but when its incr counter is used for the first time

Yes, that is the basic idea from the paper &mdash; anything that witnesses the result of an operation, either adding or removing a node, needs to ensure that the counter has been updated.  For performance reasons you want to minimize the places where/times when you ensure that the counter updates have been performed and IME it can be a little bit tricky to get right.

The proof in the paper discusses the linearization properties in more detail.

File: src_lockfree/size.ml [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+
+  let create n = Array.init n @@ fun _ -> Atomic.make collecting
+
+  let[@inline] set s i after =
+    let snap = Array.unsafe_get s i in
+    let before = Atomic.get snap in
+    if
+      before = collecting
+      || (* NOTE: The condition below accounts for overflow. *)
+      (after - before - 1) land max_value < max_value / 2
```
[the remark this answers is not in the document]
Comment (polytypic) [reply src_lockfree/size.ml:51]: The overall goal at this point (and in the similar snippet of code in `forward`) is to update the copy of a counter in the snapshot with the _latest_ value of the counter.

The counters are always only incremented.  So, typically we want to update when `after > before`. Unfortunately it is possible for a counter to overflow:
```ocaml
# Int.max_int + 1 ;;
- : int = -4611686018427387904
```
The condition above is written in such a way that the copy is (practically always) updated correctly even in case of overflow (unless the counter is incremented roughy `max_value / 2` times, which is extremely unlikely).  Here is a REPL snippet that shows results from the condition:
```ocaml
# let max_value = Int.max_int ;;
val max_value : int = 4611686018427387903
# let condition before after = (after - before - 1) land max_value < max_value / 2 ;;
val condition : int -> int -> bool = <fun>
# let plus n delta = (n + delta) land max_value ;;
val plus : int -> int -> int = <fun>
# let try_cond before delta = condition before (plus before delta) ;;
val try_cond : int -> int -> bool = <fun>
# try_cond max_value 0 ;;
- : bool = false
# try_cond max_value 1 ;;
- : bool = true
# try_cond max_value 2 ;;
- : bool = true
# try_cond max_value 3 ;;
- : bool = true
# try_cond max_value (-1) ;;
- : bool = false
# try_cond max_value (-2) ;;
- : bool = false
# try_cond max_value (-3) ;;
- : bool = false
# try_cond max_value (max_value / 2) ;;
- : bool = true
# try_cond max_value (max_value / 2 + 1) ;;
- : bool = false
# try_cond max_value (max_value / 2 - 1) ;;
- : bool = true
- : bool = true
# condition 0 (-1) ;;
- : bool = false
# condition Int.max_int (-1) ;;
- : bool = false
# condition Int.max_int (-2) ;;
- : bool = false
# condition 0 (-2) ;;
- : bool = false
```
So, basically, the update is attempted (condition is true) when the new (`after`) value seems more likely to be a new value (than an old value) and is not one the two special negative values.
